include/llbuild/Core/BuildEngine.h

```diff
@@ -101,6 +98,121 @@ struct Result {
   basic::Clock::Timestamp end;
 };
 
+
+/// An environment represents the unmodeled surroundings and side effects of
+/// tasks. In the context of a build system, for example, this could be the file
+/// system, the POSIX environment, etc. A file input task conceptually brings
+/// the file in from the environment into the llbuild computation model. The
+/// environment after a task has run may contain file system side effects (files
+/// written that were not explicitly noted as outputs).  Ideally the environment
+/// side effects should always be null, but this mechanism provides a means for
+/// non-ideal systems to still gain some of the benefits.
+///
+/// The engine stores references to environments and manages merging them on
+/// behalf of tasks executing during a build. The actual value of an
+/// environment, however, is regarded as opaque and not directly interpreted by
+/// the engine.
+typedef void* Environment;
+
+
+class TaskInterface {
```

Comment:
This is the core of the change.  Tasks now get this interface instead of a build engine reference.   Behind the scenes this will allow the engine to track environment state across each of the operations on it.